File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;

class Solution {
    public int[] solution(String[] operations) {
        int[] answer = new int[2];;
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        for(int i = 0; i < operations.length; i++){
            StringTokenizer st = new StringTokenizer(operations[i]);
            String cmd = st.nextToken();
            int n = Integer.parseInt(st.nextToken());
            if(cmd.equals("I")) {
                minHeap.add(n);
                maxHeap.add(n);
            } else {
                if(minHeap.isEmpty()) continue;
                
                if(n == 1) {
                    int max = maxHeap.poll();
                    minHeap.remove(max);
                } else {
                    int min = minHeap.poll();
                    maxHeap.remove(min);
                }
            }
        }
        
        if(minHeap.isEmpty()) {
            return new int[]{0, 0};
        } else {
            return new int[]{maxHeap.peek(), minHeap.peek()};
        }
    }
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.cpp
#include <string>
#include <vector>
#include <set>
#include <sstream>

using namespace std;

vector<int> solution(vector<string> operations) {
    vector<int> answer;
    multiset<int> ms;
    
    for(string op: operations){
        stringstream ss(op);
        char cmd;
        int num;
        ss >> cmd >> num;
        
        if(cmd == 'I'){
            ms.insert(num);
        }else if(cmd == 'D' && !ms.empty()){
            if(num == 1){
                ms.erase(--ms.end());
            }else{
                ms.erase(ms.begin());
            }
        }
    }
    if(ms.empty()){
        return {0, 0};
    }else{
        return {*ms.rbegin(), *ms.begin()};
    }
}